p4-16/spec/P4-16-spec.mdk

```diff
@@ -2717,13 +2767,64 @@ X.v1  // reference to v1
 v1    // error - v1 is not in the top-level namespace
 ~ End P4Example
 
-Similar to errors, `enum` expressions only support equality (`==`)
-and inequality (`!=`) comparisons. Expressions whose type is an `enum`
+Similar to errors, `enum` expressions without a specified backing type only support equality (`==`)
+and inequality (`!=`) comparisons. Expressions whose type is an `enum` without a specified backing type
 cannot be cast to or from any other type.
 
-Note that if an `enum` value appears in the control-plane API, the
+An `enum` that specifies a backing type, such as the following:
+
+~ Begin P4Example
+enum bit<8> E {
+  e1 = 0,
+  e2 = 1,
+  e3 = 2
+}
+~ End P4Example
+
+also supports explicit casts to and from the backing type.  For instance, the
+following code:
+
+~ Begin P4Example
+bit<8> x;
+E e1 = E.e2;
+E e2;
+
+x  = (bit<8>) e1; // sets x to 1
+e2 = (E) x;       // sets e to E.e2
+~ End P4Example
+
+casts `e1`, which was initialized to `E.e2` to a `bit<8>`, using the specified
+numeric representation for `E.e2`, `1`.  The variable `e2` is then set to the
+symbolic value `E.e2`, which corresponds to the numeric value `1`.
+
+Note that while it is always safe to cast from an `enum` to its numeric type,
+it may not be th case that every numeric value specified by the backing type
+has a related symbol in the `enum`.  For instance, the following code:
+
+~ Begin P4Example
+bit<8> x = 5;
+E e = (E) x; // sets e to an unspecified value
+~ End P4Example
+
+sets `e` to an unknown value, since there is no symbol corresponding to the
+numeric value `5`.  While the value of `e` is unspecified, it must not
+correspond to any symbol defined within enum `E`.  This is because casting from
+a numeric type to an `enum` type is required obey the following semantics:
+
+Given an enum type $E$ with underlying type `bit<W>` and the set of mappings
+of $(s, n)$ where $S$ is the set of symbols ($s$) and $N$ is the set of numeric
+representations ($n$), the following properties must hold:
+
+1. if an integer $x \notin N$ then $\forall e \in S$, `(E)x != e`
+1. $\forall x \in [0, 2^W]$, `(bit<W>)(E)x == x`
```

Comment:
I don' t know why you want this property to hold. Who cares about it?

Reply:
Do you mean to say the property holds by definition of an enum with underlying type and doesn't need to be spelled out explicitly?

Reply:
If that is the model, then we should say so explicitly in the spec.
This will preclude other implementations, and will also make it easier for people to understand what is going on. Do not leave freedom to the compiler writers if you think you will be sorry about it later. Like people designing C did with respect to integer overflow.

Reply:
This was requested in the issue #611 related to this and I basically pulled the text from there (with a lite-rewording & turning some text in math).  Where do we stand on this?

Personally, I don't feel strongly about it one way or the other.